### PR TITLE
cl_khr_external_semaphore_khr: semaphore re-import

### PR DESCRIPTION
**Pull request by lakshmih**
Add clImportSemaphoreSyncFD API call.  Sync fd semaphores must re-import the sync_fd after every wait.  Add an API call to make this re-import possible, without creating a new OpenCL semaphores.  See issue #888.

### CODE REVIEW

**Comment by bcalidas** [comment]
That makes sense. We will rename to clReimportSemaphoreSyncFdKHR.

**Comment by bcalidas** [comment]
Regarding the discussion around whether the usage clImportSemaphore API call should be required for importing an external semaphore payload  - we prefer the current approach of importing the external semaphore payload when the semaphore is created using clCreateSemaphoreWithProperties. This allows us to have a clean description of  external semaphore state soon as it is created. Our recommendation is that usage of   clImportSemapore be considered as a convenience as opposed to a requirement.

File: ext/cl_khr_external_semaphore.asciidoc
Comment (bashbaug) [inline ext/cl_khr_external_semaphore.asciidoc:330]: Discussed in the August 29th teleconference.  I think either:

1. This sentence is stale and should be updated to include all handle types (it sounded like this was the intent).
2. Or, we need to define an error code when this API is used with a non-SYNC_FD handle type.

**Review by bashbaug — COMMENTED** [review]
A few other follow-up questions about this feature:

1. What happens if there are outstanding signal or wait operations on the semaphore that is being re-imported?  Should these signal or wait operations operate on the previous semphore handle (the one that was imported when the enqueue call was made), or the new semaphore handle (the one that was imported when the command executes), or should this case be implementation-defined or undefined behavior?
2. Pending resolution of the discussion on #945, what should happen to any exported handles from a semaphore that has been re-imported?  Are they invalid?  If they are valid, do they refer to the "old" semaphore, or to the "new" semaphore?
3. I'm not sure if this needs to be documented explicitly in the spec or not, but just to confirm: calling clImportSemaphoreKHR does not modify the reference counts on the semaphore object, correct?

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:336]: Do we have any insights into what properties we may want to pass here?
May be we can fold handle_type into import_props and keep it consistent with clCreateSemaphoreWithPropertiesKHR?

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:351]: Do we need handle_type and handle_size in that case? The handle_type and size associated with sema_object should be available from sema_object itself.

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:349]: Are implementations strictly required to destroy sema_object and recreate it by associating with new handle? Or just closing earlier handle reference and associating new handle reference with sema_object good enough?

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:334]: Do we strictly need to pass handle value at the time of import if we can associate it with this new API at a later point?
We don't need to disallow, but would it make sense to make it optional or allow the handle value to be NULL?

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:332]: If the intent is to change some of the semaphore properties passed at the time of clCreateSemaphoreKHR API (associated handle_value in this case), would it make sense to rename this API to something like clSetSemaphorePropertiesKHR? May be it is easier to reuse and extend it to other cases if needed in the future?

**Review by nikhiljnv — COMMENTED** [review]
Have left some comments based on initial review.

**Comment by nikhiljnv** [comment]
> What happens if there are outstanding signal or wait operations on the semaphore that is being re-imported? Should these signal or wait operations operate on the previous semphore handle (the one that was imported when the enqueue call was made), or the new semaphore handle (the one that was imported when the command executes), or should this case be implementation-defined or undefined behavior?

+1 to Ben's comments.

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [reply ext/cl_khr_external_semaphore.asciidoc:332]: Just noting down the summary of discussion last week -
This is not a preferred path in general for OpenCL objects.
We do not want OpenCL objects to be modifiable via direct Set APIs for good reason

**Comment by bcalidas** [comment]
Regarding 
1) What happens if there are outstanding signal or wait operations on the semaphore that is being re-imported? Should these signal or wait operations operate on the previous semphore handle (the one that was imported when the enqueue call was made), or the new semaphore handle (the one that was imported when the command executes), or should this case be implementation-defined or undefined behavior?

>>> If a semaphore has a pending signal and clReleaseSemaphore is called, then the effect is likely implementation defined. We should put this clarification in the base spec.  With the latest version of the spec, the semaphore is safe to resignal/reimport as soon as WaitSemaphore returns. Therefore we only need to worry about the situation where the semaphore is released/re-imported  without WaitSemaphore ever having been called.

**Comment by bcalidas** [comment]
Regarding
2) Pending resolution of the discussion on https://github.com/KhronosGroup/OpenCL-Docs/issues/945, what should happen to any exported handles from a semaphore that has been re-imported? Are they invalid? If they are valid, do they refer to the "old" semaphore, or to the "new" semaphore?
In this situation, the semaphore logically would need to be-exported from OpenCL. Once OpenCL re-imports the semaphore handle , the externally exported handle would continue pointing to the old semaphore. Hence the need for re-export. This is also consistent with what you would expect if you released the semaphore and created a new one via import.

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:111]: Any specific reason to pass handle_ptr as a separate parameter and not make it one of the props just like supported by clCreateSemaphoreKHR?

**Comment by bcalidas** [comment]
Based on discussions around export of imported semaphores which we have decided is not permitted (  (https://github.com/KhronosGroup/OpenCL-Docs/issues/945)
and clarification of Release Semaphore Behavior ( https://github.com/KhronosGroup/OpenCL-Docs/pull/969 ) we believe that potential  behavior  issues relating to semaphore re-import have been resolved.

The next steps are to finalize the clImportSemaphoreKHR() function name and function signature.

**Comment by nikhiljnv** [comment]
As discussed in memory subgroup call on November 14, 2023, agree to move the API to cl_khr_external_semaphore_sync_fd.

**Comment by bcalidas** [comment]
Please review the latest changes. The semaphore re-import command has been moved into the sync fd extension

**Review by kpet — CHANGES_REQUESTED** [review]
This change ought to bump the extension spec number

**Comment by paulfradgley** [comment]
Looks good to me too.

**Comment by joshqti** [comment]
I pushed a commit to advanced the extension number.

**Comment by nikhiljnv** [comment]
Is the change ready to go? Or are there any reviews pending? Also, we need to resolve the conflicts on TOT.

File: ext/cl_khr_external_semaphore.asciidoc
Comment (nikhiljnv) [inline ext/cl_khr_external_semaphore.asciidoc:98]: Should this be renamed to cl_semaphore_reimport_properties_khr?
Are these going to be distinct from cl_semaphore_properties_khr that can be passed at the time of clCreateSeamphoreKHR possibly supporting initial import?

Comment (nikhiljnv) [reply ext/cl_khr_external_semaphore.asciidoc:98]: Can be merged after rename to cl_semaphore_reimport_properties_khr, version update and conflict resolution.

**Review by bashbaug — COMMENTED** [review]
LGTM, once the properties are renamed (to reimport) and the version conflict is resolved (the new version should be 0.9.2 now).  Ping me once these are fixed and I'll do a final approval and merge.

**Review by nikhiljnv — APPROVED** [review]
LGTM.

**Pull request merged**
